## Preprocessing

In [1]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import tensorflow as tf

#  Import and read the charity_data.csv.
import pandas as pd
application_df = pd.read_csv("https://static.bc-edx.com/data/dl-1-2/m21/lms/starter/charity_data.csv")
application_df.head()

EIN                                      NAME APPLICATION_TYPE  \
0  10520599              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1  10531628    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2  10547893        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3  10553066            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  10556103  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0       Independent          C1000    ProductDev   Association       1   
1       Independent          C2000  Preservation  Co-operative       1   
2  CompanySponsored          C3000    ProductDev   Association       1   
3  CompanySponsored          C2000  Preservation         Trust       1   
4       Independent          C1000     Heathcare         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

In [2]:
# Drop the non-beneficial ID columns, 'EIN' and 'NAME'.
application_df2 = application_df.drop(columns =['EIN','NAME'], axis = 1)
application_df2.head()

APPLICATION_TYPE       AFFILIATION CLASSIFICATION      USE_CASE  \
0              T10       Independent          C1000    ProductDev   
1               T3       Independent          C2000  Preservation   
2               T5  CompanySponsored          C3000    ProductDev   
3               T3  CompanySponsored          C2000  Preservation   
4               T3       Independent          C1000     Heathcare   

   ORGANIZATION  STATUS     INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  \
0   Association       1              0                      N     5000   
1  Co-operative       1         1-9999                      N   108590   
2   Association       1              0                      N     5000   
3         Trust       1    10000-24999                      N     6692   
4         Trust       1  100000-499999                      N   142590   

   IS_SUCCESSFUL  
0              1  
1              1  
2              0  
3              1  
4              1

In [3]:
# Determine the number of unique values in each column.
application_df2.nunique()

APPLICATION_TYPE            17
AFFILIATION                  6
CLASSIFICATION              71
USE_CASE                     5
ORGANIZATION                 4
STATUS                       2
INCOME_AMT                   9
SPECIAL_CONSIDERATIONS       2
ASK_AMT                   8747
IS_SUCCESSFUL                2
dtype: int64

In [4]:
# Look at APPLICATION_TYPE value counts to identify and replace with "Other"
application_df2['APPLICATION_TYPE'].value_counts()

APPLICATION_TYPE
T3     27037
T4      1542
T6      1216
T5      1173
T19     1065
T8       737
T7       725
T10      528
T9       156
T13       66
T12       27
T2        16
T25        3
T14        3
T29        2
T15        2
T17        1
Name: count, dtype: int64

In [5]:
# Choose a cutoff value and create a list of application types to be replaced
# use the variable name `application_types_to_replace`
application_types_to_replace = ['T13', 'T12', 'T2', 'T25','T14', 'T29', 'T15','T17', 'T9']

# Replace in dataframe
for app in application_types_to_replace:
    application_df['APPLICATION_TYPE'] = application_df['APPLICATION_TYPE'].replace(app,"Other")

# Check to make sure replacement was successful
application_df['APPLICATION_TYPE'].value_counts()

APPLICATION_TYPE
T3       27037
T4        1542
T6        1216
T5        1173
T19       1065
T8         737
T7         725
T10        528
Other      276
Name: count, dtype: int64

In [6]:
# Look at CLASSIFICATION value counts to identify and replace with "Other"
class_type = application_df2['CLASSIFICATION'].value_counts()
class_type

CLASSIFICATION
C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
         ...  
C4120        1
C8210        1
C2561        1
C4500        1
C2150        1
Name: count, Length: 71, dtype: int64

In [7]:
# You may find it helpful to look at CLASSIFICATION value counts >1
classifications_cleaned = class_type.loc[class_type > 1]
classifications_cleaned

CLASSIFICATION
C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
C7000      777
C1700      287
C4000      194
C5000      116
C1270      114
C2700      104
C2800       95
C7100       75
C1300       58
C1280       50
C1230       36
C1400       34
C7200       32
C2300       32
C1240       30
C8000       20
C7120       18
C1500       16
C1800       15
C6000       15
C1250       14
C8200       11
C1238       10
C1278       10
C1235        9
C1237        9
C7210        7
C2400        6
C1720        6
C4100        6
C1257        5
C1600        5
C1260        3
C2710        3
C0           3
C3200        2
C1234        2
C1246        2
C1267        2
C1256        2
Name: count, dtype: int64

In [12]:
#retrieve the classifications that are after the first 5 values in the index
classifications_to_replace= class_type.iloc[5:].index
classifications_to_replace

Index(['C7000', 'C1700', 'C4000', 'C5000', 'C1270', 'C2700', 'C2800', 'C7100',
       'C1300', 'C1280', 'C1230', 'C1400', 'C7200', 'C2300', 'C1240', 'C8000',
       'C7120', 'C1500', 'C1800', 'C6000', 'C1250', 'C8200', 'C1238', 'C1278',
       'C1235', 'C1237', 'C7210', 'C2400', 'C1720', 'C4100', 'C1257', 'C1600',
       'C1260', 'C2710', 'C0', 'C3200', 'C1234', 'C1246', 'C1267', 'C1256',
       'C2190', 'C4200', 'C2600', 'C5200', 'C1370', 'C1248', 'C6100', 'C1820',
       'C1900', 'C1236', 'C3700', 'C2570', 'C1580', 'C1245', 'C2500', 'C1570',
       'C1283', 'C2380', 'C1732', 'C1728', 'C2170', 'C4120', 'C8210', 'C2561',
       'C4500', 'C2150'],
      dtype='object', name='CLASSIFICATION')

In [13]:
# Choose a cutoff value and create a list of classifications to be replaced
# use the variable name `classifications_to_replace`
class_to_replace = list(classifications_to_replace)

# Replace in dataframe
for cls in classifications_to_replace:
    application_df['CLASSIFICATION'] = application_df['CLASSIFICATION'].replace(cls,"Other")

# Check to make sure replacement was successful
application_df['CLASSIFICATION'].value_counts()

CLASSIFICATION
C1000    17326
C2000     6074
C1200     4837
Other     2261
C3000     1918
C2100     1883
Name: count, dtype: int64

In [14]:
# Convert categorical data to numeric with `pd.get_dummies`
dummies = pd.get_dummies(application_df2, dtype=int)
dummies.head()

STATUS  ASK_AMT  IS_SUCCESSFUL  APPLICATION_TYPE_T10  APPLICATION_TYPE_T12  \
0       1     5000              1                     1                     0   
1       1   108590              1                     0                     0   
2       1     5000              0                     0                     0   
3       1     6692              1                     0                     0   
4       1   142590              1                     0                     0   

   APPLICATION_TYPE_T13  APPLICATION_TYPE_T14  APPLICATION_TYPE_T15  \
0                     0                     0                     0   
1                     0                     0                     0   
2                     0                     0                     0   
3                     0                     0                     0   
4                     0                     0                     0   

   APPLICATION_TYPE_T17  APPLICATION_TYPE_T19  ...  INCOME_AMT_1-9999  \
0                     0                     0  ...                  0   
1                     0                     0  ...                  1   
2                     0                     0  ...                  0   
3                     0                     0  ...                  0   
4                     0                     0  ...                  0   

   INCOME_AMT_10000-24999  INCOME_AMT_100000-499999  INCOME_AMT_10M-50M  \
0                       0                         0                   0   
1                       0                         0                   0   
2                       0                         0                   0   
3                       1                         0                   0   
4                       0                         1                   0   

   INCOME_AMT_1M-5M  INCOME_AMT_25000-99999  INCOME_AMT_50M+  \
0                 0                       0                0   
1                 0                       0                0   
2                 0                       0                0   
3                 0                       0                0   
4                 0                       0                0   

   INCOME_AMT_5M-10M  SPECIAL_CONSIDERATIONS_N  SPECIAL_CONSIDERATIONS_Y  
0                  0                         1                         0  
1                  0                         1                         0  
2                  0                         1                         0  
3                  0                         1                         0  
4                  0                         1                         0  

[5 rows x 117 columns]

In [15]:
# Split our preprocessed data into our features and target arrays
y = dummies['IS_SUCCESSFUL'].values
X= dummies.drop(columns = 'IS_SUCCESSFUL').values

# Split the preprocessed data into a training and testing dataset
X_train, X_test, y_train, y_test = train_test_split(X,y)

In [16]:
# Create a StandardScaler instances
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

## Compile, Train and Evaluate the Model

In [18]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
input_total = len(X_train[0])
hidden_nodes_layer1 = 3
hidden_nodes_layer2= 5
nn = tf.keras.models.Sequential()

# First hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer1, input_dim = input_total,activation = 'relu'))

# Second hidden layer
nn.add(tf.keras.layers.Dense(units= hidden_nodes_layer2, activation='relu'))

# Output layer
nn.add(tf.keras.layers.Dense(units=1, activation='sigmoid'))

# Check the structure of the model
nn.summary()

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_3 (Dense)             (None, 3)                 351       
                                                                 
 dense_4 (Dense)             (None, 5)                 20        
                                                                 
 dense_5 (Dense)             (None, 1)                 6         
                                                                 
Total params: 377 (1.47 KB)
Trainable params: 377 (1.47 KB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


In [19]:
# Compile the model
nn.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

In [20]:
# Train the model
fit_model = nn.fit(X_train_scaled, y_train, epochs = 100)

Epoch 1/100
804/804 [==============================] - 3s 2ms/step - loss: 0.6464 - accuracy: 0.6556
Epoch 2/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5959 - accuracy: 0.7145
Epoch 3/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5784 - accuracy: 0.7177
Epoch 4/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5700 - accuracy: 0.7231
Epoch 5/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5665 - accuracy: 0.7245
Epoch 6/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5645 - accuracy: 0.7252
Epoch 7/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5627 - accuracy: 0.7255
Epoch 8/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5612 - accuracy: 0.7279
Epoch 9/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5601 - accuracy: 0.7271
Epoch 10/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5596 - accura

In [21]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 0s - loss: 0.5653 - accuracy: 0.7227 - 488ms/epoch - 2ms/step
Loss: 0.5652987957000732, Accuracy: 0.7226822376251221


In [22]:
# Export our model to HDF5 file
filepath = r'C:\Users\pegam\deep-learning-challenge\HDF5 Files'
nn.save(filepath, save_format = 'h5')

/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py:3103: UserWarning: You are saving your model as an HDF5 file via `model.save()`. This file format is considered legacy. We recommend using instead the native Keras format, e.g. `model.save('my_model.keras')`.
  saving_api.save_model(


Optimize the model

In [23]:
#Test 1

# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
input_total = len(X_train[0])
hidden_nodes_layer1 = 10
hidden_nodes_layer2= 10
nn = tf.keras.models.Sequential()

# First hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer1, input_dim = input_total,activation = 'relu'))
# Second hidden layer
nn.add(tf.keras.layers.Dense(units= hidden_nodes_layer2, activation='relu'))
# Output layer
nn.add(tf.keras.layers.Dense(units=1, activation='sigmoid'))
# Check the structure of the model
nn.summary()
# Compile the model
nn.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
# Train the model
fit_model = nn.fit(X_train_scaled, y_train, epochs = 100)

Model: "sequential_2"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_6 (Dense)             (None, 10)                1170      
                                                                 
 dense_7 (Dense)             (None, 10)                110       
                                                                 
 dense_8 (Dense)             (None, 1)                 11        
                                                                 
Total params: 1291 (5.04 KB)
Trainable params: 1291 (5.04 KB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________
Epoch 1/100
804/804 [==============================] - 2s 2ms/step - loss: 0.6063 - accuracy: 0.7052
Epoch 2/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5645 - accuracy: 0.7299
Epoch 3/100
804/804 [==============================] - 3s 4ms/step - loss: 0.5571 - 

In [24]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 0s - loss: 0.5697 - accuracy: 0.7262 - 432ms/epoch - 2ms/step
Loss: 0.569696843624115, Accuracy: 0.7261807322502136


In [25]:
#Test 2

# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
input_total = len(X_train[0])
hidden_nodes_layer1 = 15
hidden_nodes_layer2= 10
hidden_nodes_layer3= 15
nn = tf.keras.models.Sequential()

# First hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer1, input_dim = input_total,activation = 'relu'))
# Second hidden layer
nn.add(tf.keras.layers.Dense(units= hidden_nodes_layer2, activation='relu'))
nn.add(tf.keras.layers.Dense(units= hidden_nodes_layer3, activation='relu'))

# Output layer
nn.add(tf.keras.layers.Dense(units=1, activation='sigmoid'))
# Check the structure of the model
nn.summary()
# Compile the model
nn.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
# Train the model
fit_model = nn.fit(X_train_scaled, y_train, epochs = 100)

Model: "sequential_3"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_9 (Dense)             (None, 15)                1755      
                                                                 
 dense_10 (Dense)            (None, 10)                160       
                                                                 
 dense_11 (Dense)            (None, 15)                165       
                                                                 
 dense_12 (Dense)            (None, 1)                 16        
                                                                 
Total params: 2096 (8.19 KB)
Trainable params: 2096 (8.19 KB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________
Epoch 1/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5993 - accuracy: 0.6921
Epoch 2/100
804/804 [==============================] 

In [26]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 1s - loss: 0.5666 - accuracy: 0.7275 - 675ms/epoch - 3ms/step
Loss: 0.5666221976280212, Accuracy: 0.7274635434150696


In [27]:
#Test 3

# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
input_total = len(X_train[0])
hidden_nodes_layer1 = 7
hidden_nodes_layer2= 14
hidden_nodes_layer3 = 21
nn = tf.keras.models.Sequential()

# First hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer1, input_dim = input_total,activation = 'relu'))
# Second hidden layer
nn.add(tf.keras.layers.Dense(units= hidden_nodes_layer2, activation='relu'))
#third hidden later
nn.add(tf.keras.layers.Dense(units= hidden_nodes_layer3, activation='relu'))

# Output layer
nn.add(tf.keras.layers.Dense(units=1, activation='sigmoid'))
# Check the structure of the model
nn.summary()
# Compile the model
nn.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
# Train the model
fit_model = nn.fit(X_train_scaled, y_train, epochs = 200)

Model: "sequential_4"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_13 (Dense)            (None, 7)                 819       
                                                                 
 dense_14 (Dense)            (None, 14)                112       
                                                                 
 dense_15 (Dense)            (None, 21)                315       
                                                                 
 dense_16 (Dense)            (None, 1)                 22        
                                                                 
Total params: 1268 (4.95 KB)
Trainable params: 1268 (4.95 KB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________
Epoch 1/200
804/804 [==============================] - 3s 2ms/step - loss: 0.6128 - accuracy: 0.6976
Epoch 2/200
804/804 [==============================] 

In [30]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 0s - loss: 0.5694 - accuracy: 0.7276 - 463ms/epoch - 2ms/step
Loss: 0.5693594217300415, Accuracy: 0.727580189704895


In [29]:
#save file
filepath = r'C:\Users\pegam\deep-learning-challenge\HDF5 Files'
nn.save(filepath, save_format = 'h5')

Analysis Overview of the Analysis
The model aims to forecast successful applicants to help Alphabet Soup in better selecting candidates for funding.

Results:
Data Preprocessing
What variable(s) are the target(s) for your model?

The target variable was the project's success status.

What variable(s) are the features for your model?

Feature variables encompassed all other attributes.

What variable(s) should be removed from the input data because they are neither targets nor features?

Variables such as EIN and Name were excluded as they did not serve as targets or features.

Compiling, Training, and Evaluating the Model
How many neurons, layers, and activation functions did you select for your neural network model, and why?

Initially, two layers were chosen, mirroring our previous work during data preprocessing. Subsequently, I increased it to three layers to explore potential performance enhancement.

Were you able to achieve the target model performance?

No, the desired model performance was not attained.

What steps did you take in your attempts to increase model performance?

Attempts to enhance model performance included adding an extra layer, adjusting the number of hidden nodes, and varying the number of epochs.

Summary
Overall, the model utilized did not achieve optimization. The initial model came closest with an accuracy of 73%, indicating that increasing layers and epochs did not necessarily augment accuracy in this instance.

Future Recommendations
Based on these findings, future explorations could focus on alternative model architectures or feature engineering techniques to improve performance.